### PR TITLE
[IMP] base: add limit when filtered records

### PR DESCRIPTION
**Pull request by fmdl**
Hi guys

This commit is to improve speed in some filtered operations.
- when you need to select only one record, it is better to stop the loop when you have find one. Replace `self.filtered(func)[:1]` by `self.filtered(func, limit=1)`.
- when you use filtered to get a boolean : 
```
if self.filtered(func):
      return ....
```
you don't need to compute the condition for all records, you can add `limit=1` and `count=True`

what do you think ?

@odony 
@alnslang 
@Yenthe666 

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Elkasitu** [comment]
Do you have some numbers on what the performance gain would be from this?

Also, not a big fan of the `count` parameter as setting it changes the output type of the filtered method, it's inconsistent compared to `search` and `search_count`

**Comment by fmdl** [comment]
For exemple here (I have not check all the odoo code, to find some place to use the limit arg): https://github.com/odoo/odoo/blob/12.0/addons/stock/models/stock_picking.py#L405
it don't need to test all record, and don't need to return a new recordset.

I agree with you about the `count`, that I havmade two commit.

About performance it depends of the number of record and the time to check.

**Comment by mreficent** [comment]
I think that doing 
```python
if self.filtered(func, limit=1):
      return ...
```
is the same that doing
```python
if self.filtered(func, limit=1, count=True):
      return ...
```
So adding the `count` is not a real gain.

**Comment by fmdl** [comment]
@mreficent `count=True` remove the browse().

I agree the `count` is not a real gain, I can remove that.

**Comment by odony** [comment]
> For exemple here (I have not check all the odoo code, to find some place to use the limit arg): https://github.com/odoo/odoo/blob/12.0/addons/stock/models/stock_picking.py#L405

Unless I'm mistaken, it would make no measurable difference for that specific example. Because of the prefetching, the cost of the filter function is the same whether you call it once or _n_ times (`ml.result_package_id` causes only 1 SQL query for all moves lines)
I suspect similar conclusions could be reached in many other examples where `filtered()` is used: prefetching will nullify any benefit of a "limit", and the rest of the cost is negligible (instantiating records or iterating records are many orders of magnitude cheaper than a single roundtrip to the database!)

This means we'd be introducing a `limit` parameter that looks like it will make a difference in the amount of records processed (like SQL's LIMIT), while in reality it changes almost nothing. I find this greatly misleading.

Simply put, `.filtered()` should never be used in performance-sensitive places, because it is by definition not efficient - it operates on data in memory. If performance is a concern, the data should rather *not* have been fetched from the database in the first place: use search() properly!
I see so many horrors where devs are using `model.search(domain).filtered(func)` instead of using a proper `domain`, that I'd rather we improve the API to **discourage** this, rather than **encouraging** it with a `limit` parameter.

**Comment by fmdl** [comment]
I understand your point of view.

At the begining, I have work on it to replace `record.filtered(func)[:1]`, by a function `filtered_one(func)` `= filtered(func, limit=1)`.

if there more than 1000 records, there are some benefit, but it is not the most commun case.

In replacement of self[0] or self[:1], we could create `def first_one(self)` and `def last_one(self)`, what do you think ?
